Enhance the benchmark runner to support multiple top level objects use case.

cheqianh opened the pull request:
### Description:
This PR adds support for multiple top level use case, so that we have an apple-to-apple performance comparison for Ion/JSON/CBOR. 

### Note:
- This PR focuses on `io_type=file`, we might consider adding `io_type=buffer` later but we need to figure out how to fairly use JSON and CBOR's load/dump APIs to iterate each of multiple top level object - https://github.com/amazon-ion/ion-python/issues/325.

- When we benchmark read speed for JSON document that includes multiple top level objects, we suppose that the given data is in JSONL. So we have to modify our JSON sample data - [code here](https://github.com/amazon-ion/ion-python/pull/315/files#diff-6d077184aeda689a4e665e62c8cc7c7bb2f8d33073523539f42a3a94d26b55b2R1)

- I'm not sure if protobuf has the same issue for the multiple top-level objects use case. This need to be figure out when we come back to protobuf in the future - https://github.com/amazon-ion/ion-python/issues/326.

- Three files I used for test locally, I upload them for visible but they will not be included when merge this PR into the main branch. They are:
1 [amazon/ionbenchmark/json_newline_vs_list_repro.py](https://github.com/amazon-ion/ion-python/pull/315/files#diff-061791b385cd9f2ffb42daa826b3896f5765ec9d2ff8182a89058d7f24672669)
2 [amazon/ionbenchmark/format_conversion_correct.py](https://github.com/amazon-ion/ion-python/pull/315/files#diff-b70ae6cd92930d856e7b4de45eff0422f5e468ffd7e8decea9af49f659562bef)
3 [amazon/ionbenchmark/format_conversion_inaccurate.py](https://github.com/amazon-ion/ion-python/pull/315/files#diff-402e60208fa4e59d092a876473c7ca3390d4f406c7b5a3293e7fc7851a66ba5c)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

tgregg commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -114,21 +114,51 @@ def test_fn():
                     return loader_dumper.dump(data_obj, f)
+        elif _format.format_is_binary(data_format) or _format.format_is_ion(data_format):
+            def test_fn():
+                with tempfile.TemporaryFile(mode="ab") as f:
+                    for top_level_obj in data_obj:
+                        loader_dumper.dump(top_level_obj, f)
```
For Ion, is this going to write a new symbol table for each top-level value? That would not lead to the most efficient encoding. How does the performance and data size for Ion compare when using this technique vs using a single dump of the entire list of top-level values?

cheqianh replied:
```
loader_dumper.dump(data_obj, f, sequence_as_stream=True)
| (ion_binary,dumps,service_log_legacy) |       22263038 |  2027258750.00 |   2035399545.00 |               42571303 |
```
```
loader_dumper.dump(data_obj, f, sequence_as_stream=False)
| (ion_binary,dumps,service_log_legacy) |       22263038 |  2010238558.40 |   2024798020.82 |               42571457 |
```
```
for top_level_obj in data_obj:
    loader_dumper.dump(top_level_obj, f)
| (ion_binary,dumps,service_log_legacy) |       22263038 |  2486109233.40 |   2495525362.50 |                  44581 |
```

Changed to `loader_dumper.dump(data_obj, f, sequence_as_stream=True)`

tgregg commented on amazon/ionbenchmark/benchmark_spec.py:
```diff
@@ -176,12 +176,37 @@ def get_input_file_size(self):
+                            break
+                    self._data_object = cbor_objects
+            elif format_is_protobuf(format_option):
+                with open(self.get_input_file(), "rb") as fp:
+                    self._data_object = loader.load(fp)
```
If we're not sure whether the protobuf logic is correct, then I'd rather throw an error for now.

cheqianh replied:
Changed.

tgregg commented on amazon/ionbenchmark/format_conversion_inaccurate.py:
```diff
@@ -0,0 +1,42 @@
+# Flag to control if the JSON file is delimited by a newline.
+generate_jsonl = True
+
+with open(ion_file, 'br') as fp:
```
For posterity, can you reply here to explain what was wrong with this technique?

cheqianh replied:
Sure, the reasons are:
1. Since JSON doesn't support multiple top-level objects, we used to store all JSON object within a list. Now we iterate each top level values and write them one by one. 
2. CBOR conversion was wrong. We used jsonable bytes for JSON/CBOR conversion instead of Python objects that are converted from JSON objects so that the resulted CBOR data is incorrect.

tgregg commented on tests/test_benchmark_cli.py:
```diff
@@ -218,3 +219,16 @@ def test_format_conversion_ion_text_to_ion_binary():
+)
+def test_multiple_top_level_values(args):
+    (command, format_option, file) = args
+    (error_code, _, _) = run_cli([f'{command}', file, '--format', f'{format_option}', '--io-type', 'file'])
+    assert not error_code
```
Are we able to assert that multiple top-level values are actually read/written?

cheqianh replied:
As discussed offline, I'll create a few new tests for each generated test_fun to ensure they work correctly.

tgregg commented on tests/test_benchmark_spec.py:
```diff
@@ -11,6 +14,51 @@ def _generate_test_path(p):
+            jsonl = f.readline()
+            if jsonl == '':
+                break
+            # make sure the json object are equivalence
```
```suggestion
            # make sure the json object are equivalent
```

cheqianh replied:
Changed.

tgregg commented on tests/test_benchmark_spec.py:
```diff
@@ -11,6 +14,51 @@ def _generate_test_path(p):
+        while True:
+            try:
+                o = cbor2.load(f)
+                # make sure the CBOR object are equivalence
```
```suggestion
                # make sure the CBOR object are equivalent
```

cheqianh replied:
Changed.

cheqianh commented:
File `multiple_top_level_object.json` and `multiple_top_level_object.cbor` includes equivalent data:
```
{"name":"John", "age":30, "car":null}
{"name":"Mike", "age":33, "car":null}
{"name":"Jack", "age":24, "car":null}
```

rmarrowstone commented on amazon/ionbenchmark/benchmark_spec.py:
```diff
@@ -176,12 +176,37 @@ def get_input_file_size(self):
             loader = self.get_loader_dumper()
-            with open(self.get_input_file(), "rb") as fp:
-                self._data_object = loader.load(fp)
+            format_option = self.get_format()
```
Why have all this branching here? we already have `loader`s for each type. Put the loader specific code there.

cheqianh replied:
Okay, I'll refactor them into a specific file there. I need to run the benchmark-cli again to ensure that the write/read process won't affect the performance (e.g., values generated by yield instead of returning an iterator directly).

cheqianh replied:
Refactor Json and Cbor, the performance are almost the same but the memory usage peak have a signaficant difference, I'm going to see why. 

Also since we dump the whole document for Ion to avoid unnecessary symbol table write, the Ion_load_dump needs to be changed too.

rmarrowstone replied:
I wonder if it is due to building the list?

rmarrowstone replied:
I'm also not sure why IonLoadDump needs to change. My understanding is that we're not flushing the writer so it will just buffer until it closes, then flush the symbol table and values. Right?

cheqianh replied:
Re:
> I'm also not sure why IonLoadDump needs to change. My understanding is that we're not flushing the writer so it will just buffer until it closes, then flush the symbol table and values. Right?

I saw a 20% difference between dumps each top-level object and dump the whole document. I'll try to see why it takes shorter when repeatedly calling dump

rmarrowstone commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -114,21 +114,51 @@ def test_fn():
-            with open(data_file, "rb") as f:
-                return loader_dumper.load(f)
+            format_option = benchmark_spec.get_format()
+            if _format.format_is_ion(format_option):
```
Put the loader specific logic in custom loaders for each type, avoid branching here.

cheqianh commented:
The commit [74d89ad](https://github.com/amazon-ion/ion-python/pull/315/commits/74d89ad7630434a10dad8dc50071ca9bb9713167) refactor loader (read APIs) to separate files. And below are the new results for a log
| name                                 |   file_size(B) |   time_min(ns) |   time_mean(ns) |   memory_usage_peak(B) |
| ------------------------------------- | --------------- | --------------- | ---------------- | ----------------------- |
| Ion - after                          |       22263038 |  5092575416.60 |   5109052881 |                 276500 |
| Ion - before                         |       22263038 |                |   4730782811    |                 171700 |
||
| json - after |      145330385 |   871989650.00 |    872934369 |                 112139 |
| json - before |      145330385 |   |    875,974,056 |                 77,446 |
||
| cbor - after |      116960762 |   985745625.00 |    991288347 |                  87769 |
| cbor - before |      116960762 |   |    995,595,161 |                  51,222 |

cheqianh commented:
Below is the table showing the differences in write performance after the refactor.

| name                                 |   file_size(B) |   time_min(ns) |   time_mean(ns) |   memory_usage_peak(B) |
| ------------------------------------- | --------------- | --------------- | ---------------- | ----------------------- |
| Ion - before                         |       22263038 |               |   2,509,895,173    |                 42,573,352 |
| Ion - after                          |       22263038 |  |   4,139,845,225 |               42,573,352  |
||
| json - before |      145330385 |   |    3,981,317,292 |                 143,755,789 |
| json - after |      145330385 |   |   4,018,459,501 |              143,756,301   |
||
| cbor - before |      116960762 |   |    1,303,162,089 |                  19,301 |
| cbor - after |      116960762 |   |   1,291,955,078 |                 19,813 |

tgregg commented:
Why is Ion read and write performance worse after the refactor?

rmarrowstone commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -112,23 +116,90 @@ def test_fn():
+                                      "support yet.")
 
     elif match_arg == ['file', 'write', 'load_dump']:
+        # This method should return a list that holds all top_level values.
```
Why? Why not return a generator?

cheqianh replied:
Okay yeah, we can reuse the `loader.load` for that. Working on this.

rmarrowstone commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -112,23 +116,90 @@ def test_fn():
+            if custom_file:
+                def test_fn():
+                    with open(custom_file, 'ab') as f:
+                        loader_dumper.dump(data_obj, f, sequence_as_stream=True)
```
The purpose of the `loader_dumper` abstraction is to encapsulate format stuff like `sequence_as_stream`. Why does that need to be passed here?

cheqianh replied:
Changed

rmarrowstone commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -112,23 +116,90 @@ def test_fn():
+            # Refer to https://github.com/amazon-ion/ion-python/issues/326
+            raise NotImplementedError("Benchmarking Protocol Buffer multiple top level object use case may not "
+                                      "support yet.")
+        elif _format.format_is_cbor(data_format):
```
There's a lot of duplicated code here. There are three things that vary:
1. custom file or not.
2. binary or text.
3. ion specific flag.
1 and 2 should be pretty easily dealt with by defining what to open before opening it. 3 should be dealt with by putting ion specific stuff in the ion loader dumper.

cheqianh replied:
Got it. I had all branches separated for debugging reasons to allow me to navigate the specific formats affected by the refactor. I will include the refactor for these three types in the next commit.

cheqianh commented:
Re, 
> I'm also not sure why IonLoadDump needs to change. My understanding is that we're not flushing the writer so it will just buffer until it closes, then flush the symbol table and values. Right?

I saw a 20% difference between `dumps` each top-level object and `dump` the whole document. I can take a look at that later.

cheqianh commented:
The commit [2b177af](https://github.com/amazon-ion/ion-python/pull/315/commits/2b177af8b85f9a82e49de9aaba9ccc19e79cdb35) addressed the comment. A few comments and highlights below. 
1. Re: https://github.com/amazon-ion/ion-python/pull/315#discussion_r1442404936
It seems that the generator can't be fully consumed by test_fun; it will only parse the first value and return when I pass a generator to it. I refactored the `benchmark_spec.get_data_object()` method to return a generator and reused it in other places to avoid repeating code. However, for benchmarking, we still call `list(the_generator)` to convert it into a list before passing it to `test_fun`. Do you have any concerns or comments about this?

2. Re: https://github.com/amazon-ion/ion-python/pull/315#discussion_r1442408471
I'm going to investigate why this happens and determine which benchmarking methods are more accurate. Based on the previous results, there is a 20% difference between these two approaches.

3. After the refactor, the Ion write performance dropped from 2e^9 to 4e^9 I'm not sure why; I'm investigating it now.

4. There are some conflicts with the main branch. (I guess it's because of the recent bare_value PR) I'll work on them at the end.

rmarrowstone commented:
>    It seems that the generator can't be fully consumed by test_fun; it will only parse the first value and return when I pass a generator to a defined `test_fun`. I refactored the `benchmark_spec.get_data_object()` method to return a generator and reused it in other places to avoid repeating code. However, for benchmarking, we still call `list(the_generator)` to convert it into a list before passing it to `test_fun`. Do you have any concerns or comments about this?

Can't we just pull off the generator and do nothing with the values in the `test_fun`?

cheqianh commented:
The highlight No.3 in my [comment](https://github.com/amazon-ion/ion-python/pull/315#issuecomment-1881885164) above is solved and I updated the metrics table for dump APIs.

This is because for the `before` metrics, I checked out the latest ion-python commit to see how much improvement we have made for the write side. However, since the PR is not merged yet so that the `after` result is based on the latest released version 0.11.3. I updated it. In the future I will use this commit [d425587](https://github.com/amazon-ion/ion-python/commit/d425587d94f8e9fa22368ed1b82a1b65fcb81820) for benchmarking comparison in this PR (the current latest commit).

cheqianh commented:
Investigating the build failure issue, it worked well locally.

rmarrowstone commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -84,19 +84,23 @@ def run_benchmark(benchmark_spec: BenchmarkSpec):
     """
     Create a benchmark function for the given `benchmark_spec`.
+
+    :param return_obj: If the test_fun returns the load object for debugging. It only works for `io-type=file` and
```
I don't think we're using this pydoc style anywhere else. simpleion uses the google pydoc style. That is the main API to this code, follow what that does.

cheqianh replied:
Changed.

rmarrowstone commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -105,30 +109,64 @@ def test_fn():
             return loader_dumper.loads(buffer)
 
     elif match_arg == ['buffer', 'write', 'load_dump']:
+        # This method returns a list
```
remove?

cheqianh replied:
Removed.

rmarrowstone commented on amazon/ionbenchmark/benchmark_spec.py:
```diff
@@ -177,12 +181,23 @@ def get_input_file_size(self):
+            else:
+                rtn = []
+                with open(read_file, 'br' if format_is_bytes(format_option) else 'r') as fp:
+                    obj = loader.load(fp)
+                    for v in obj:
+                        rtn.append(v)
```
```suggestion
                    rtn = [v for v in loader.load(fp)]
```

Shorter, and I believe, faster.

cheqianh replied:
Changed, thanks for catching this up.

tgregg commented on amazon/ionbenchmark/ion_load_dump.py:
```diff
@@ -26,12 +26,17 @@ def loads(self, s):
     def dumps(self, obj):
         ion.c_ext = self._c_ext
-        return ion.dumps(obj, binary=self._binary)
+        ion.dumps(obj, binary=self._binary)
```
Should this use `sequence_as_stream=True`?

cheqianh replied:
io_type=buffer currently doesn't handle multiple top-level use case, I opened an issue for that - https://github.com/amazon-ion/ion-python/issues/325

rmarrowstone commented on amazon/ionbenchmark/cbor_load_dump.py:
```diff
@@ -0,0 +1,27 @@
+import cbor2
+
+
+class CborLoadDump:
```
`Cbor2LoadDump`?

cheqianh replied:
changed.

rmarrowstone commented on amazon/ionbenchmark/ion_load_dump.py:
```diff
@@ -26,12 +26,17 @@ def loads(self, s):
     def dumps(self, obj):
         ion.c_ext = self._c_ext
-        return ion.dumps(obj, binary=self._binary)
+        ion.dumps(obj, binary=self._binary)
```
does this need to change too?

cheqianh replied:
Added back.

rmarrowstone replied:
The question was as to whether this needed `sequence_as_stream=True`

rmarrowstone commented on amazon/ionbenchmark/json_load_dump.py:
```diff
@@ -0,0 +1,27 @@
+            yield json.loads(line)
+
+    def loads(self, s):
+        return json.loads(s)
```
so loads can only handle single values?

cheqianh replied:
`io_type=buffer` currently doesn't handle multiple top-level use case, I opened an issue for that - https://github.com/amazon-ion/ion-python/issues/325

rmarrowstone replied:
Ok. At first blush it seems like just making multiple TLVs there would be simpler than creating an issue and having divergent code paths, but we need to get this merged.

cheqianh commented on amazon/ionbenchmark/Format.py:
```diff
@@ -25,11 +25,21 @@ def format_is_cbor(format_option):
            or (format_option == Format.PROTOBUF.value) or (format_option == Format.SD_PROTOBUF.value)
 
 
+def format_is_bytes(format_option):
```
`ion_text` is not a binary format, but it needs a 'b' flag to open the file.

rmarrowstone replied:
I understand the logic requirement. What I don't understand is what 'is_binary' is actually called for other than determining the flag.

cheqianh commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -105,30 +109,64 @@ def test_fn():
+            if custom_file:
+                if _format.format_is_bytes(data_format):
+                    def test_fn():
+                        with open(custom_file, 'ab') as f:
```
Changed. Yeah it works, thanks for the recommendation.

cheqianh commented on amazon/ionbenchmark/benchmark_runner.py:
```diff
@@ -105,30 +109,64 @@ def test_fn():
+        data_obj = list(data_obj)
 
+        data_format = benchmark_spec.get_format()
+        if _format.format_is_protobuf(data_format):
```
I'm not sure if protobuf can or should handle multiple top-level objects. I opened an issue for this - https://github.com/amazon-ion/ion-python/issues/326 and throw an error for now JIC.
